src/Resources/contao/library/Contao/User.php

```diff
@@ -258,478 +277,425 @@ public function getData()
+	 * {@inheritdoc}
 	 */
-	protected function regenerateSessionId()
+	public function setSalt($salt)
 	{
-		$container = \System::getContainer();
-		$strategy = $container->getParameter('security.authentication.session_strategy.strategy');
+		$this->salt = $salt;
 
-		// Regenerate the session ID to harden against session fixation attacks
-		switch ($strategy)
-		{
-			case SessionAuthenticationStrategy::NONE:
-				break;
+		return $this;
+	}
 
-			case SessionAuthenticationStrategy::MIGRATE:
-				$container->get('session')->migrate(false); // do not destroy the old session
-				break;
 
-			case SessionAuthenticationStrategy::INVALIDATE:
-				$container->get('session')->invalidate();
-				break;
+	/**
+	 * {@inheritdoc}
+	 */
+	public function isAccountNonExpired()
+	{
+		$time = time();
 
-			default:
-				throw new \RuntimeException(sprintf('Invalid session authentication strategy "%s"', $strategy));
-		}
+		return ($this->start == '' || $this->start < $time) && ($this->stop == '' || $this->stop > $time);
 	}
 
 
 	/**
-	 * Generate a session
+	 * {@inheritdoc}
 	 */
-	protected function generateSession()
+	public function isAccountNonLocked()
 	{
 		$time = time();
 
-		// Generate the cookie hash
-		$this->strHash = $this->getSessionHash($this->strCookie);
-
-		// Clean up old sessions
-		$this->Database->prepare("DELETE FROM tl_session WHERE tstamp<? OR hash=?")
-					   ->execute(($time - \Config::get('sessionTimeout')), $this->strHash);
+		return ($this->locked + \Config::get('lockPeriod')) < $time;
+	}
 
-		// Save the session in the database
-		$this->Database->prepare("INSERT INTO tl_session (pid, tstamp, name, sessionID, ip, hash) VALUES (?, ?, ?, ?, ?, ?)")
-					   ->execute($this->intId, $time, $this->strCookie, \System::getContainer()->get('session')->getId(), $this->strIp, $this->strHash);
 
-		// Set the authentication cookie
-		$this->setCookie($this->strCookie, $this->strHash, ($time + \Config::get('sessionTimeout')), null, null, \Environment::get('ssl'), true);
+	/**
+	 * {@inheritdoc}
+	 */
+	public function isCredentialsNonExpired()
+	{
+		return true;
 	}
 
 
 	/**
-	 * Remove the authentication cookie and destroy the current session
-	 *
-	 * @return boolean True if the user could be logged out
+	 * {@inheritdoc}
 	 */
-	public function logout()
+	public function isEnabled()
 	{
-		// Return if the user has been logged out already
-		if (!\Input::cookie($this->strCookie))
-		{
-			return false;
-		}
 
-		$intUserid = null;
+		return ! (bool) ($this->disable);
+	}
 
-		// Find the session
-		$objSession = $this->Database->prepare("SELECT * FROM tl_session WHERE hash=?")
-									 ->limit(1)
-									 ->execute($this->strHash);
 
-		if ($objSession->numRows)
-		{
-			$this->strIp = $objSession->ip;
-			$this->strHash = $objSession->hash;
-			$intUserid = $objSession->pid;
-		}
+	/**
+	 * {@inheritdoc}
+	 */
+	public function serialize()
+	{
+		return serialize(array(
+			$this->id,
+			$this->username,
```

Comment:
I think we need to save `tstamp` here as well so we can use the timestamp to compare if the `refreshUser` returned a new object than was in the security token.

Reply:
Yeah, thats something still missing (see open task in the list).